tests/models/test_process_rules.py

```diff
@@ -1,38 +1,214 @@
 import pytest
+import mock
+
+import pandas
 from dmscripts.models.process_rules import (
     format_datetime_string_as_date,
     remove_username_from_email_address,
-    extract_id_from_user_info
+    extract_id_from_user_info,
+    query_data_from_config,
+    process_data_from_config
 )
+from dmapiclient import DataAPIClient
+
+
+class TestFormatDatetimeStringAsDate:
+
+    def test_format_datetime_string_as_date(self):
+        initial_date = "2016-10-08T12:00:00.00000Z"
+        formatted_date = "2016-10-08"
+        assert format_datetime_string_as_date(initial_date) == formatted_date
+
+    def test_format_datetime_string_as_date_raises_error_if_initial_date_format_incorrect(self):
+        initial_dates = (
+            "2016-10-08T12:00:00.00000",
+            "2016-10-08T12:00:00",
+            "2016-10-08"
+        )
+        for date in initial_dates:
+            with pytest.raises(ValueError) as excinfo:
+                format_datetime_string_as_date(date)
+
+            assert "time data '{}' does not match format".format(date) in str(excinfo.value)
+
+
+class TestRemoveUsernameFromEmailAddress:
+
+    def test_remove_username_from_email_address(self):
+        initial_email_address = "user.name@domain.com"
+        formatted_email_address = "domain.com"
+        assert remove_username_from_email_address(initial_email_address) == formatted_email_address
+
+
+class TestExtractIDFromUserInfo:
+
+    def test_extract_id_from_user_info(self):
+        user_list = [{'id': x} for x in range(3)]
+        extracted_ids = extract_id_from_user_info(user_list)
+
+        assert extracted_ids == '0,1,2'
+
+
+class TestQueryDataFromConfig:
+
+    @mock.patch('dmscripts.models.queries.base_model')
+    def test_query_data_from_config_queries_db_given_a_base_model(self, base_model):
+        config = {
+            'name': 'completed_outcomes',
+            'base_model': 'outcomes',
+            'get_data_kwargs': {'completed': True},
+            'keys': (
+                'id',
+                'result',
+                ('resultOfDirectAward', 'project', 'id',),
+            )
+        }
+        client = mock.Mock(autospec=DataAPIClient)
+        expected_query_result = pandas.DataFrame([{'id': '1', 'result': 'awarded'}])
+        base_model.side_effect = [expected_query_result]
+
+        assert query_data_from_config(
+            config, logger=mock.ANY, limit=100, client=client, output_dir="data"
+        ).equals(expected_query_result)
+
+        # What this test is really concerned with - we call queries.base_model rather than queries.model
```

Comment:
Do any of the tests care about what methods were called on `client`?

Reply:
Those calls are buried in `ModelTrawler` - `test_modeltrawler.py` has coverage there.

Reply:
Sounds like a joy.